### ISSUE TITLE
The MissingParameterTypeHint sniff does not detect union array-based types

### ISSUE THREAD

**Issue by morozov** [slevomat/coding-standard#856]
Consider the code example:
```php
<?php

declare(strict_types=1);

/**
 * @param array<string> $array
 */
function acceptArray1($array) : void
{
}

/**
 * @param array<string>|array<int> $array
 */
function acceptArray2($array) : void
{
}
```
Running `phpcs` with the Doctrine Coding Standard (tested with slevomat/coding-standard 6.1.0) produces the following:
```
$ phpcs test.php

------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
------------------------------------------------------------------------------------------------
 8 | ERROR | [x] Function acceptArray1() does not have parameter type hint for its parameter
   |       |     $array but it should be possible to add it based on @param annotation
   |       |     "array<string>".
   |       |     (SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingParameterTypeHint)
------------------------------------------------------------------------------------------------
```
The sniff detects `acceptArray1()` as missing a type hint but doesn't detect `acceptArray2()`. However, the `array` type hint is applicable to it as well since `array` is a common denominator of its all union type components.

**Comment by kukulich** [comment]
Thank you for your report.

Fixed in https://github.com/slevomat/coding-standard/commit/c22cf74d00dfa31c00fc9a628843d68c867bb9aa

**Comment by morozov** [comment]
Hm… it seems like the problem wasn't resolved completely. Running `phpcs` with the Doctrine Coding Standard and slevomat/coding-standard 6.1.1, I'm getting the following results:
```
$ composer show | grep slevomat/coding-standard
slevomat/coding-standard                       6.1.1   Slevomat Coding Standard for PHP_...

$ phpcs test.php

-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 8 | ERROR | [x] Function acceptArray1() does not have native type hint for its parameter
   |       |     $array but it should be possible to add it based on @param annotation
   |       |     "array<string>".
   |       |     (SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingNativeTypeHint)
-----------------------------------------------------------------------------------------------
```
The name of the violation is now different ("MissingNativeTypeHint" instead of "MissingParameterTypeHint") but `acceptArray2()` is still not flagged.

For some reason, when processing the second method, instead of the `ArrayTypeNode`, something detected as `GenericTypeNode` in: https://github.com/slevomat/coding-standard/blob/b54110e048f3c5c64534b82650976a7eeba1880f/SlevomatCodingStandard/Sniffs/TypeHints/ParameterTypeHintSniff.php#L200-L202

```
$ composer show | grep phpdoc-parser
phpstan/phpdoc-parser                          0.3.5   PHPDoc parser with support...
```

The real code that I used the issue (https://github.com/doctrine/dbal/pull/3839) is also not flagged.

**Comment by kukulich** [comment]
> `The name of the violation is now different ("MissingNativeTypeHint" instead of "MissingParameterTypeHint")`

It's ok. There were changes in 6.0.0 - TypeHintDeclaration was splitted to more sniffs.

> For some reason, when processing the second method, instead of the ArrayTypeNode, something detected as GenericTypeNode

`array<string>` is `GenericTypeNode`. `string[]` would be `ArrayTypeNode`.

I really don't understand the bug know. I test lowest dependencies in travis and `phpstan/phpdoc-parser` 0.3.5 is ok there :(

**Comment by morozov** [comment]
> `array<string>` is `GenericTypeNode`. `string[]` would be `ArrayTypeNode`.

So if `array<string>` is not detected as `ArrayTypeNode` but the condition above expects an `ArrayTypeNode`, isn't it where the problem is? However, according to the test, it should be detected too.

https://github.com/slevomat/coding-standard/blob/c22cf74d00dfa31c00fc9a628843d68c867bb9aa/tests/Sniffs/TypeHints/data/parameterTypeHintErrors.fixed.php#L149-L154

Is there any additional information I could provide or I just debug it deeper?

**Comment by kukulich** [comment]
@morozov It would be perfect if you could debug it deeper (eg. prepare failing test). I cannot reproduce it now.

The condition on line 200 should not be used here. This should be solved in condition on line 152.

**Comment by morozov** [comment]
Before I debug, could you please run the following and confirm that `phpcs` flags the missing type declaration for you:
```
cd /tmp
git clone git@github.com:morozov/dbal.git --depth 1 --branch slevomat/coding-standard/issues/856
cd dbal
composer install
vendor/bin/phpcs lib/Doctrine/DBAL/Connection.php
```
In my case, `phpcs` doesn't find any issues in this file:
```
. 1 / 1 (100%)


Time: 2.63 secs; Memory: 18MB
```

**Comment by kukulich** [comment]
@morozov I can reproduce it so I will fix it.

**Comment by kukulich** [comment]
Should be fixed in https://github.com/slevomat/coding-standard/commit/65ecb08e9d0ab964362892564b38d8a3a392d9f6

@morozov Can you test dev-master please?

**Comment by morozov** [comment]
Now the sniff works as expected:
```
$ phpcs lib/Doctrine/DBAL/Connection.php

FILE: lib/Doctrine/DBAL/Connection.php
------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
------------------------------------------------------------------------------------------------
 788 | ERROR | [x] Method \Doctrine\DBAL\Connection::executeQuery() does not have native type
     |       |     hint for its parameter $types but it should be possible to add it based on
     |       |     @param annotation "array<int, (int|string)>|array<string,
     |       |     (int|string)>".
     |       |     (SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingNativeTypeHint)
------------------------------------------------------------------------------------------------
```
The fixer also works:
```diff
diff --git a/lib/Doctrine/DBAL/Connection.php b/lib/Doctrine/DBAL/Connection.php
index 054e59b81..a5bd9c160 100644
--- a/lib/Doctrine/DBAL/Connection.php
+++ b/lib/Doctrine/DBAL/Connection.php
@@ -785,7 +785,7 @@ class Connection implements DriverConnection
      *
      * @throws DBALException
      */
-    public function executeQuery(string $query, array $params = [], $types = [], ?QueryCacheProfile $qcp = null) : ResultStatement
+    public function executeQuery(string $query, array $params = [], array $types = [], ?QueryCacheProfile $qcp = null) : ResultStatement
     {
         if ($qcp !== null) {
             return $this->executeCacheQuery($query, $params, $types, $qcp);
```
Thank you!
